fvtest/compilertriltest/LinkageTest.cpp

```diff
@@ -0,0 +1,136 @@
+
+template<> const char * TypeToString<int32_t>::prefix = "i"; 
+template<> const char * TypeToString<int64_t>::prefix = "l";
+template<> const char * TypeToString<float>  ::prefix = "f";
+template<> const char * TypeToString<double> ::prefix = "d"; 
+
+TEST(TypeToString,TestTemplate) { 
+   EXPECT_STREQ("Int32", TypeToString<int32_t>::type);
+   EXPECT_STREQ("Double", TypeToString<double>::type);
+}
+
+
+template <typename T> 
+T passThrough(T x) { return x; } 
+
+
+template <typename T> 
+class LinkageTest : public TRTest::JitTest {}; 
+
+typedef ::testing::Types<int32_t, int64_t, float, double> InputTypes;
+TYPED_TEST_CASE(LinkageTest, InputTypes);
+
+TYPED_TEST(LinkageTest, SystemLinkageParameterPassingSingleArg) { 
+    char inputTrees[200] = {0};
+    const auto format_string = "(method return=%s args=[%s] (block (%sreturn (%scall address=%p args=[%s] linkage=system (%sload parm=0)) )  ))";
+    std::snprintf(inputTrees, 200, format_string, TypeToString<TypeParam>::type, // Return
+                                                  TypeToString<TypeParam>::type, //Args
+                                                  TypeToString<TypeParam>::prefix, //return
+                                                  TypeToString<TypeParam>::prefix, //call
+                                                  &passThrough<TypeParam>, //address
+                                                  TypeToString<TypeParam>::type, //args 
+                                                  TypeToString<TypeParam>::prefix //load
+                                                  );
+    auto trees = parseString(inputTrees);
+
+    ASSERT_NOTNULL(trees) << "Trees failed to parse\n" << inputTrees;
+
+    Tril::DefaultCompiler compiler{trees};
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+
+    auto entry_point = compiler.getEntryPoint<TypeParam (*)(TypeParam)>();
+
+    EXPECT_EQ(static_cast<TypeParam>(0), entry_point( static_cast<TypeParam>(0)))  << "Input Trees: " << inputTrees;
+    EXPECT_EQ(static_cast<TypeParam>(1), entry_point( static_cast<TypeParam>(1)))  << "Input Trees: " << inputTrees;
+    EXPECT_EQ(static_cast<TypeParam>(-1), entry_point(static_cast<TypeParam>(-1))) << "Input Trees: " << inputTrees;
+}
+
+template <typename T> 
+T fourthArg(T a, T b, T c, T d) { return d; } 
+
+
+TYPED_TEST(LinkageTest, SystemLinkageParameterPassingFourArg) { 
+    char inputTrees[400] = {0};
+    const auto format_string = "(method return=%s args=[%s,%s,%s,%s] (block (%sreturn (%scall address=%p args=[%s,%s,%s,%s] linkage=system"
+                                 " (%sload parm=0)"
+                                 " (%sload parm=1)"
+                                 " (%sload parm=2)"
+                                 " (%sload parm=3)"
+                                 ") )  ))";
+
+    std::snprintf(inputTrees, 400, format_string, TypeToString<TypeParam>::type,   // Return
```

Comment:
This might be the biggest argument in favour of exposing a type-genericized builder pattern for the Tril AST. 

My thought being something like 

```
return<T>().
  call<T>().
    load<T>(0);
```

Reply:
FYI I think this is also causing compilation failures on older compiler versions. More here: https://github.com/eclipse/omr/issues/2152